bin/mango-notebook

```diff
@@ -17,59 +17,149 @@
 # limitations under the License.
 #
 
-set -ex
+shopt -s extglob
+## Global variable declarations
+# Bash 4.3 introduce nameref feature which able to write bash
+# script without global variable.
+# Unfortunately some major distribution use bash 4.2 such as centOS 7.
 
+declare -a SPARK_ARGS=()
+declare -a NOTEBOOK_ARGS=()
 
-SOURCE_DIR=$(dirname ${BASH_SOURCE[0]})
+err_report() {
+  local -ir line=$1
+  local -ir exit_code=$2
+  echo "Error on line ${line} with exit code ${exit_code}"
+}
 
-# Split args into Spark and ADAM args
-DD=False  # DD is "double dash"
-PRE_DD=()
-POST_DD=()
-for ARG in "$@"; do
-  shift
-  if [[ $ARG == "--" ]]; then
-    DD=True
-    POST_DD=( "$@" )
-    break
+trap 'err_report $LINENO' ERR
+
+print_err(){
+  local -r msg="$1"
+  printf 'ERROR: %s\n'  "${msg}" 1>&2
+}
+
+# Retrieve the full path of a program using its name. If the program is not
+# found using $PATH viariable the script exit on error
+#
+# Usage program_exists [name] {true|false}
+# Argument 1 program name
+# Argument 2 optional boolean flag to exit or not if the program is not found
+#            default is true (exit on error)
+# Return full path or empty string if not found
+get_program_path(){
+  local -r name="$1"
+  local -r exit_on_error=${2:-true}
+  local -r path=$(command -v "${name}")
+  if [[ -z "${path}" ]]; then
+    if ${exit_on_error} ; then
+      print_err "${name} is not found."
+      exit 1
+    else
+      print_err "${name} is not found; Aborting."
+    fi
   fi
-  PRE_DD+=("$ARG")
-done
+  echo ${path}
+}
 
-if [[ $DD == True ]]; then
-  SPARK_ARGS=("${PRE_DD[@]}")
-  NOTEBOOK_ARGS=("${POST_DD[@]}")
-else
-  SPARK_ARGS=()
-  NOTEBOOK_ARGS=("${PRE_DD[@]}")
-fi
+# Retrieve the path of py4j zip archive used by current spark
+get_py4jzip(){
+  local -a py4j_zip_all_versions
+  local py4j_zip=""
+  if [[ -d "${SPARK_HOME}/python/lib" ]]; then
+    py4j_zip_all_versions=( ${SPARK_HOME}/python/lib/py4j-+([0-9])\.+([0-9])?(\.+([0-9]))-src.zip )
+    # use the latest version
+    py4j_zip="${py4j_zip_all_versions[-1]}"
+    if [[ ! -e ${py4j_zip} ]]; then
+      err_report ${LINENO} $? "${py4j_zip} is not found; Aborting."
```

Comment:
When this error is found, it only exits out of the function, but the script succeeds. This failure should end the script.